### PR TITLE
GDScript: Add module description in markdown

### PR DESCRIPTION
**Pull request by anvilfolk**
This adds a .md file to the GDScript module that gives a moderately in-depth description of how the module is organized, used, main classes, etc :)

### CODE REVIEW

File: modules/gdscript/README.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+
```
Comment (YuriSizov) [inline modules/gdscript/README.md:1]: Is this empty line accidental?

Comment (anvilfolk) [reply modules/gdscript/README.md:1]: Nope, really wasn't! :)

File: modules/gdscript/README.md
[the remark this answers is not in the document]
Comment (anvilfolk) [reply modules/gdscript/README.md:1]: This makes sense somewhere in the contributing to the engine part, right?

File: modules/gdscript/README.md (hunk 1 above)
Comment (YuriSizov) [reply modules/gdscript/README.md:1]: You mean we need an empty line for some reason?

File: modules/gdscript/README.md
[the remark this answers is not in the document]
Comment (anvilfolk) [reply modules/gdscript/README.md:1]: Oh, I understand now! So something like "If you're looking to learn more about GDScript, please see here", and that links to https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/index.html?

File: modules/gdscript/README.md (hunk 1 above)
Comment (anvilfolk) [reply modules/gdscript/README.md:1]: Oh my goodness, my bad! I meant it really wasn't _intentional_, not _accidental_ 😅

File: modules/gdscript/README.md
[the remark this answers is not in the document]
Comment (YuriSizov) [reply modules/gdscript/README.md:1]: Maybe something like "You can learn more about GDScript in the [documentation](https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/index.html). It describes the syntax and user facing systems and concepts, and can be used as a reference for what user expectations are."

File: modules/gdscript/README.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+This provides some basic information in how GDScript is implemented and integrates with the rest of the engine. You can learn more about GDScript in the [documentation](https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/index.html). It describes the syntax and user facing systems and concepts, and can be used as a reference for what user expectations are.
+
+
+## General design
+
+GDScript is:
+
+1. A dynamically typed language. Type hints help with static analysis and performance, but typed code must easily interoperate with untyped code.
+2. A tightly designed language. Features are added because they are _needed_, and not because they can be added or are interesting to develop.
+3. Primarily an interpreted scripting language. It is meant to be easy to use and develop gameplay in. It is not meant for CPU-intensive algorithms or data processing, and is not optimized for it. For that, C# or GDExtensions may be used.
```
Comment (adamscott) [inline modules/gdscript/README.md:11]:
```suggestion
3. Primarily an interpreted scripting language. It is meant to be easy to use and develop gameplay in. It is not meant for CPU-intensive algorithms or data processing, and is not optimized for it. For that, C# or [GDExtension](https://docs.godotengine.org/en/stable/tutorials/scripting/gdextension/what_is_gdextension.html) may be used.
```

File: modules/gdscript/README.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+
+## Compilation
+Scripts can be at different stages of compilation, but the process generally consists of a linear series of steps: tokenizing, parsing, analyzing, and finally compiling.
+
+The main class of the GDScript module is the [`GDScript`](gdscript.h) class, which represents a GDScript file, and contains information regarding all that that file contains, such as main class and references to subclasses, members, signals, implicit methods like initializers, etc. This is the main class that the compilation step deals with.
```
Comment (adamscott) [inline modules/gdscript/README.md:17]:
```suggestion
The main class of the GDScript module is the [`GDScript`](gdscript.h) class, which usually represents a GDScript file, and contains information regarding all that that file contains, such as main class and references to subclasses, members, signals, implicit methods like initializers, etc. This is the main class that the compilation step deals with.

```note
Each subclass are themselves `GDScript` instances.
```
```

File: modules/gdscript/README.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+### Tokenizing (see [`gdscript_tokenizer.h/cpp`](gdscript_tokenizer.h))
+
+Tokenizing is the process of converting the source code `String` into a sequence of tokens, which represent language constructs, identifiers, literals, etc. It happens as the first step of the parsing process, and so isn't really found independently of that in the codebase, save very rare exceptions.
+
+### Parsing (see [`gdscript_parser.h/cpp`](gdscript_parser.h))
+
+The parser takes a sequence of tokens and builds the abstract syntax tree (AST) of the GDScript program. This is what finds any syntax errors in a GDScript program. The AST is used in the analyzing and compilation steps, and the source code `String` and sequence of tokens are discarded. 
+
+The parser class also defines all the possible nodes of the syntax tree, e.g. `GDScriptParser::ClassNode` for a node representing a GDScript class. It further defines the typechecking datastructures like `GDScriptParser::Datatype`.
```
Comment (dalexeev) [inline modules/gdscript/README.md:35]: It may be worth mentioning the parser limitations that we [deliberately added](https://godotengine.org/article/gdscript-progress-report-writing-new-parser/#less-lookahead). The parser makes **one** pass: unlike the analyzer, some nodes have not yet been parsed. At any given time, only 2 tokens are available to the parser: the previous and the current. Understanding these facts has helped me in several PRs when I was trying to do something that was impossible.

File: modules/gdscript/README.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
+# Basic GDScript module architecture
```
Comment (dalexeev) [inline modules/gdscript/README.md:1]: Inconsistent blank lines in the file. I suggest using 2 blank lines between sections and 1 blank line to separate the body from the header.

```md
## Header

Text.


## Next header

Text.
```

File: modules/gdscript/README.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+The `GDScript` class contains all the information related to the corresponding GDScript class: its name and path, its members like variables, functions, symbols, signals, implicit methods like initializers, etc. This is the main class that the compilation step deals with.
+
+A secondary class is `GDScriptInstance`, defined in the same file, containing _runtime_ information for an instance of a `GDScript`, and is more related to the execution of a script by the virtual machine.
+
+### Loading source code
+
+This mostly happens by calling `GDScript::load_source_code()` on a `GDScript` object. Parsing only requires a `String`, so it is entirely possible to parse a script without a `GDScript` object!
+
+### Tokenizing (see [`gdscript_tokenizer.h/cpp`](gdscript_tokenizer.h))
```
Comment (dalexeev) [inline modules/gdscript/README.md:27]:
```suggestion
### Tokenizing (see [`GDScriptTokenizer`](gdscript_tokenizer.h))
```

File: modules/gdscript/README.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+Tokenizing is the process of converting the source code `String` into a sequence of tokens, which represent language constructs, identifiers, literals, etc. It happens as the first step of the parsing process, and so isn't really found independently of that in the codebase, save very rare exceptions.
+
+### Parsing (see [`gdscript_parser.h/cpp`](gdscript_parser.h))
```
Comment (dalexeev) [inline modules/gdscript/README.md:31]:
```suggestion
### Parsing (see [`GDScriptParser`](gdscript_parser.h))
```

File: modules/gdscript/README.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+The parser takes a sequence of tokens and builds the abstract syntax tree (AST) of the GDScript program. This is what finds any syntax errors in a GDScript program. The AST is used in the analyzing and compilation steps, and the source code `String` and sequence of tokens are discarded. 
+
+The parser class also defines all the possible nodes of the syntax tree, e.g. `GDScriptParser::ClassNode` for a node representing a GDScript class. It further defines the typechecking datastructures like `GDScriptParser::Datatype`.
+
+### Analysis and typechecking (see [`gdscript_analyzer.h/cpp`](gdscript_analyzer.h))
```
Comment (dalexeev) [inline modules/gdscript/README.md:37]:
```suggestion
### Analysis and typechecking (see [`GDScriptAnalyzer`](gdscript_analyzer.h))
```

**Review by dalexeev — COMMENTED** [review]
Great job! I didn’t know much of this or wasn’t completely sure. I've added some suggestions, you don't have to follow them all, these are just ideas that might inspire you.

The README is intended for new contributors, so it's quite difficult to decide which details are important to understand the module and which will require learning the code anyway, so it's not worth the effort.

In my opinion, a section about tests would help new contributors. There is a [doc page](https://docs.godotengine.org/en/stable/contributing/development/core_and_modules/unit_testing.html), but it is mainly about `doctest`, not about GDScript tests. It would be nice to briefly explain that tests should be positive and negative, how our files are located (folders, `*.gd`, `*.notest.gd`, `*.out`), how to run tests locally.

I hope Adam, Dmitry and George will add their suggestions too.

**Comment by anvilfolk** [comment]
@dalexeev excellent comments! I will try to find some time soon to incorporate them! I think a section about tests makes a lot of sense, and we can link to the official docs about it!

File: modules/gdscript/README.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+Scripts can be at different stages of compilation, but the process generally consists of a linear series of steps: tokenizing, parsing, analyzing, and finally compiling.
+
+The main class of the GDScript module is the [`GDScript`](gdscript.h) class, which represents a class defined in GDScript. Each `.gd` file is called a _class file_ because it implicitly defines a class in GDScript, and thus results in an associated `GDScript` object. However, GDScript classes may define [_inner classes_](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_basics.html#inner-classes), and those are also represented by further `GDScript` objects, even though they are not in files of their own.
+
+The `GDScript` class contains all the information related to the corresponding GDScript class: its name and path, its members like variables, functions, symbols, signals, implicit methods like initializers, etc. This is the main class that the compilation step deals with.
+
+A secondary class is `GDScriptInstance`, defined in the same file, containing _runtime_ information for an instance of a `GDScript`, and is more related to the execution of a script by the virtual machine.
```
[the remark this answers is not in the document]
Comment (anvilfolk) [reply modules/gdscript/README.md:21]: I opted away from mentioning `PropertyInfo/MethodInfo`, as I think that's getting a little into the weeds of the analyzer, but added a section on how GDScript interfaces with Godot, using the classes you mention!

File: modules/gdscript/README.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+### Loading source code
+
+This mostly happens by calling `GDScript::load_source_code()` on a `GDScript` object. Parsing only requires a `String`, so it is entirely possible to parse a script without a `GDScript` object!
+
+### Tokenizing (see [`gdscript_tokenizer.h/cpp`](gdscript_tokenizer.h))
+
+Tokenizing is the process of converting the source code `String` into a sequence of tokens, which represent language constructs, identifiers, literals, etc. It happens as the first step of the parsing process, and so isn't really found independently of that in the codebase, save very rare exceptions.
```
[the remark this answers is not in the document]
Comment (anvilfolk) [reply modules/gdscript/README.md:29]: I've actually very rarely touched the tokenizer and parser, so I trust your expertise here. What I noticed is that references to the tokenizer only ever shows up in the 1) parser, 2) gdscript editor, 3) highlighter, and 4) test framework. I think you're right though, the highlighter doesn't really seem to use it a lot, just references some tokens.

[the remark this answers is not in the document]
Comment (anvilfolk) [reply modules/gdscript/README.md:29]: I rewrote this a little bit, because the exception was meant to say that we don't see it independent of the parser, except in e.g. the code editor (which is one of the rare exceptions) :)

File: modules/gdscript/README.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+## Special types of scripts
+
+Certain types of GDScripts behave slightly differently. For example, autoloads are loaded with `ResourceLoader::load()` during `Main::start()`, very soon after Godot is launched. Many systems aren't initialized at that time, so error reporting is often significantly reduced and may not even show up in the editor.
+
+Tool scripts, declared with the `@tool` annotation on a GDScript file, run in the editor itself as opposed to just when the game is launched. This leads to a significant increase in complexity, as many things that can be changed in the editor may affect a currently executing tool script.
+
+
+## Other
+
+There are many other classes in the GDScript module. Here is a brief overview of some of them:
```
[the remark this answers is not in the document]
Comment (anvilfolk) [reply modules/gdscript/README.md:85]: Yes, I had forgotten about these, but I agree: the readme isn't meant to be fully comprehensive, but merely a starting point for folks wanting to get into basic GDScript work. If they want to get into those advanced topics, they will already be familiar enough with this that they can independently go figure those out :)

File: modules/gdscript/README.md (hunk 4 above)
Comment (anvilfolk) [reply modules/gdscript/README.md:35]: Yes!!!! I totally forgot that that was a super intentional design decision that is _absolutely_ worth mentioning here!

File: modules/gdscript/README.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+The `GDScript` class contains all the information related to the corresponding GDScript class: its name and path, its members like variables, functions, symbols, signals, implicit methods like initializers, etc. This is the main class that the compilation step deals with.
+
+A secondary class is `GDScriptInstance`, defined in the same file, containing _runtime_ information for an instance of a `GDScript`, and is more related to the execution of a script by the virtual machine.
+
+### Loading source code
+
+This mostly happens by calling `GDScript::load_source_code()` on a `GDScript` object. Parsing only requires a `String`, so it is entirely possible to parse a script without a `GDScript` object!
```
[the remark this answers is not in the document]
Comment (anvilfolk) [reply modules/gdscript/README.md:25]: I think this is a good idea! Though we need to remember to change it if the feature (which is hotly requested...) ever does get implemented :)

See https://github.com/godotengine/godot-proposals/issues/4220 and recent discussions in https://github.com/godotengine/godot-proposals/issues/6031

File: modules/gdscript/README.md [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+GDScript is:
+
+1. A [gradually typed](https://en.wikipedia.org/wiki/Gradual_typing) language. Type hints are optional and help with static analysis and performance. However, typed code must easily interoperate with untyped code.
+2. A tightly designed language. Features are added because they are _needed_, and not because they can be added or are interesting to develop.
+3. Primarily an interpreted scripting language: it is compiled to GDScript byte code and interpreted in a GDScript virtual machine. It is meant to be easy to use and develop gameplay in. It is not meant for CPU-intensive algorithms or data processing, and is not optimized for it. For that, [C#](https://docs.godotengine.org/en/stable/tutorials/scripting/c_sharp/c_sharp_basics.html) or [GDExtension](https://docs.godotengine.org/en/stable/tutorials/scripting/gdextension/what_is_gdextension.html) may be used.
+
+
+## Integration into Godot
+
+GDScript is integrated into godot as a module. Since modules are optional, this means that Godot may be built without GDScript and work perfectly fine without it!
```
Comment (akien-mga) [inline modules/gdscript/README.md:16]:
```suggestion
GDScript is integrated into Godot as a module. Since modules are optional, this means that Godot may be built without GDScript and work perfectly fine without it!
```

Comment (anvilfolk) [reply modules/gdscript/README.md]: Oh, nice catch! Thanks, and fixed :)

File: modules/gdscript/README.md [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+
+
+## Integration into Godot
+
+GDScript is integrated into Godot as a module. Since modules are optional, this means that Godot may be built without GDScript and work perfectly fine without it!
+
+The GDScript module interfaces with Godot's codebase by inheriting from Godot's scripting-related classes. New languages inherit from [`ScriptLanguage`](/core/object/script_language.h), and are registered in Godot's [`ScriptServer`](/core/object/script_language.h). Scripts, referring to a file containing code, inherit from `Script`. Instances of that script, which are used at runtime when actually executing the code, inherit from [`ScriptInstance`](/core/object/script_instance.h).
+
+To access Godot's internal classes, GDScript uses [`ClassDB`](/core/object/class_db.h). `ClassDB` is where Godot registers classes, methods and properties that it wants exposed to its scripting system. This is how GDScript understands what a `Vector3` is, that `print()` exists, or that `Node` has a `get_parent()` method.
```
Comment (akien-mga) [inline modules/gdscript/README.md:20]: I'm not sure the last sentence is accurate. Isn't ClassDB only for Object-derived classes?
I believe both Variant types and GDScript methods are registered via a different mechanism.

Should be confirmed, I'm no expert.

File: modules/gdscript/README.md [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+### Tokenizing (see [`GDScriptTokenizer`](gdscript_tokenizer.h))
+
+Tokenizing is the process of converting the source code `String` into a sequence of tokens, which represent language constructs (such as `for` or `if`), identifiers, literals, etc. This happens almost exclusively during the parsing process, which asks for the next token in order to make sense of the source code. The tokenizer is only used outside of the parsing process in very rare exceptions.
+
+
+### Parsing (see [`GDScriptParser`](gdscript_parser.h))
+
+The parser takes a sequence of tokens and builds [the abstract syntax tree (AST)](https://en.wikipedia.org/wiki/Abstract_syntax_tree) of the GDScript program. The AST is used in the analyzing and compilation steps, and the source code `String` and sequence of tokens are discarded. This is the process that finds any syntax errors in a GDScript program and reports them to the user. 
+
+The parser class also defines all the possible nodes of the AST, e.g. `GDScriptParser::ClassNode` for a node representing a GDScript class, or `GDScriptParser::FunctionNode` for a function. It further defines typechecking datastructures like `GDScriptParser::Datatype`.
```
Comment (akien-mga) [inline modules/gdscript/README.md:48]: Is it worth clarifying that "node" in this context is unrelated to the Node class?

Comment (akien-mga) [inline modules/gdscript/README.md:48]:
```suggestion
The parser class also defines all the possible nodes of the AST, e.g. `GDScriptParser::ClassNode` for a node representing a GDScript class, or `GDScriptParser::FunctionNode` for a function. It further defines typechecking data structures like `GDScriptParser::Datatype`.
```

File: modules/gdscript/README.md [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+
+1. Class interface resolution: to know what is available from one class, you don't need to analyze it in its entirety. You may merely analyze its "interface", meaning its name and its members (and their types!), as well as name and members of its base case, recursively. Details can be found in `GDScriptAnalyzer::resolve_class_interface()`.
+2. Out of order member resolution: the analyzer may not need an entire class interface to be resolved in order to figure out a specific type. For example, if class `A` contains code that references `B.is_alive`, then the analyzer doesn't need to immediately resolve `B`'s interface. It may simply check whether `is_alive` exists in `B`. This pattern happens often in the analyzer, and can be easily identified by the declaration of a temporary datatype with `GDScriptParser::DataType resolving_datatype;`, followed by `resolving_datatype.kind = GDScriptParser::DataType::RESOLVING;`. If the analyzer attempts to resolve a member that is already resolving, then a cyclic dependency problem has been found.
```
Comment (akien-mga) [inline modules/gdscript/README.md:67]: Up for discussion, I can see both being accepted. The class is `DataType` so two words, but the local itself is `resolving_datatype`.
```suggestion
2. Out of order member resolution: the analyzer may not need an entire class interface to be resolved in order to figure out a specific type. For example, if class `A` contains code that references `B.is_alive`, then the analyzer doesn't need to immediately resolve `B`'s interface. It may simply check whether `is_alive` exists in `B`. This pattern happens often in the analyzer, and can be easily identified by the declaration of a temporary data type with `GDScriptParser::DataType resolving_datatype;`, followed by `resolving_datatype.kind = GDScriptParser::DataType::RESOLVING;`. If the analyzer attempts to resolve a member that is already resolving, then a cyclic dependency problem has been found.
```

File: modules/gdscript/README.md [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+
+Importantly, the compilation process of a class, specifically `GDScriptCompiler::_compile_class()` _cannot_ depend on the compilation process of another class, since that will result in cyclic dependency problems. Supported cyclic dependencies must have been fully resolved during the analyze step. It is OK for supporting methods, such as `GDScriptCompiler::_populate_class_methods()` to be called on other classes as long as those methods rely exclusively on information from previous steps.
+
+
+## Loading scripts
+
+GDScripts can be loaded in a couple of different ways. The main method that the Godot itself uses to load scripts is through the `ResourceLoader` singleton. In this way, GDScripts are resources like any others, and `ResourceLoader::load()` will route to `ResourceFormatLoaderGDScript::load()` found in `gdscript.h/cpp`(gdscript.h). This generates a GDScript object which is compiled and ready to use.
```
Comment (akien-mga) [inline modules/gdscript/README.md:79]:
```suggestion
GDScripts can be loaded in a couple of different ways. The main method that Godot itself uses to load scripts is through the `ResourceLoader` singleton. In this way, GDScripts are resources like any others, and `ResourceLoader::load()` will route to `ResourceFormatLoaderGDScript::load()` found in [`gdscript.h/cpp`](gdscript.h). This generates a GDScript object which is compiled and ready to use.
```

File: modules/gdscript/README.md [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+
+The other method is to manually load the source code, then pass it to a parser, then to an analyzer and then to a compiler. This is what happens behind the scenes of the previous approach, minus some caching of GDScripts. It is used in the [GDScript test runner infrastructure](tests/gdscript_test_runner.h`).
```
Comment (akien-mga) [inline modules/gdscript/README.md:81]:
```suggestion
The other method is to manually load the source code, then pass it to a parser, then to an analyzer and then to a compiler. This is what happens behind the scenes of the previous approach, minus some caching of GDScripts. It is used in the [GDScript test runner infrastructure](tests/gdscript_test_runner.h).
```

File: modules/gdscript/README.md [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+
+
+## Other
+
+There are many other classes in the GDScript module. Here is a brief overview of some of them:
+
+- Declaration of GDScript warnings in [`GDScriptWarning`](gdscript_warning.h).
+- [`GDScriptFunction`](gdscript_function.h), which represents an executable GDScript function. The relevant file contains both static as well as runtime information.
+- The [virtual machine](gdscript_vm.cpp) is essentially defined as calling `GDScriptFunction::call()`.
+- Editor related functions can be found in parts of `GDScriptLanguage`, originally declared in [`gdscript.h`](gdscript.h) but defined in [`gdscript_editor.cpp`](gdscript_editor.cpp). Code highlighting can be found in [`GDScriptSyntaxHighlighter`](editor/gdscript_highlighter.h).
```
Comment (akien-mga) [inline modules/gdscript/README.md:109]:
```suggestion
- Editor-related functions can be found in parts of `GDScriptLanguage`, originally declared in [`gdscript.h`](gdscript.h) but defined in [`gdscript_editor.cpp`](gdscript_editor.cpp). Code highlighting can be found in [`GDScriptSyntaxHighlighter`](editor/gdscript_highlighter.h).
```

**Review by akien-mga — COMMENTED** [review]
Amazing write-up, I learned a lot :)

File: modules/gdscript/README.md [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+3. Primarily an interpreted scripting language: it is compiled to GDScript byte code and interpreted in a GDScript virtual machine. It is meant to be easy to use and develop gameplay in. It is not meant for CPU-intensive algorithms or data processing, and is not optimized for it. For that, [C#](https://docs.godotengine.org/en/stable/tutorials/scripting/c_sharp/c_sharp_basics.html) or [GDExtension](https://docs.godotengine.org/en/stable/tutorials/scripting/gdextension/what_is_gdextension.html) may be used.
+
+
+## Integration into Godot
+
+GDScript is integrated into godot as a module. Since modules are optional, this means that Godot may be built without GDScript and work perfectly fine without it!
+
+The GDScript module interfaces with Godot's codebase by inheriting from Godot's scripting-related classes. New languages inherit from [`ScriptLanguage`](/core/object/script_language.h), and are registered in Godot's [`ScriptServer`](/core/object/script_language.h). Scripts, referring to a file containing code, inherit from `Script`. Instances of that script, which are used at runtime when actually executing the code, inherit from [`ScriptInstance`](/core/object/script_instance.h).
```
[the remark this answers is not in the document]
Comment (anvilfolk) [reply modules/gdscript/README.md:18]: Adding these changes, except for mentioning the "resource" type. It's a concept they might not know/understand just yet and which isn't super relevant for immediately understanding GDScript - might be a case of less is more here? :)

File: modules/gdscript/README.md (hunk 16 above)
Comment (anvilfolk) [reply modules/gdscript/README.md:67]: If it's OK, I think I prefer the way it is right now since like you mention it is `resolving_datatype` and not `resolving_data_type`.  This way we don't have to worry about writing data type vs datatype depending on context!

**Comment by anvilfolk** [comment]
> Amazing write-up, I learned a lot :)

Thanks, it's all I'm capable of doing at the moment, but it sounds like folks might find it helpful, so that feels nice :)

It might eventually make sense to add a section on how unit tests are run, but that requires digging more deeply into that code, which I haven't done in a while :)

**Comment by anvilfolk** [comment]
Spent a few more hours on this today. I think it's good to merge if folks don't find anything else they'd like changed.

Changelog:
- Fixed some references to `_populate_class_members()` which no longer exists and has been replaced with `_prepare_compilation()`
- Added a lot more information and examples to the analyzer, since it had very, very little information
- Added more information about how cyclic dependencies are resolved
- Added some information about how typing leads to optimizations during compilation
- Tiny edits here and there

I think it's helpful for this document to describe a few of the things the analyzer actually does. This means that new contributors don't need to work through a couple hundred lines of code to try to figure it out - they can read it and get a general idea for what it does, and then read the code with more context in mind.

**Comment by Mickeon** [comment]
May I poke this? It looks great to me. Would be nice to have sooner than later for the developers

**Review by dalexeev — APPROVED** [review]
Looks good to me. Of course, there is always more to add, but we are not trying to write exhaustive internal documentation. This is a good overview for new contributors, and we can edit it later anyway (unless @vnen, @adamscott, @vonagam, and other contributors have suggestions right now). @anvilfolk Thank you again for the great write-up!

**Comment by akien-mga** [comment]
Thanks! Great writeup :)
